##### **Simplest XAI method - Permutation Feature Importance**

In [2]:
import matplotlib.pyplot as plt
import seaborn as sns 
import numpy as np 
import pandas as pd
import xgboost as xgb 
from sklearn.metrics import accuracy_score, confusion_matrix

In [3]:
credit_score = pd.read_csv("data/credit_score.csv")

In [6]:
# Select features 
features = ['INCOME', 'DEBT', 'R_EXPENDITURE', 'R_ENTERTAINMENT', 'CAT_GAMBLING']
X = credit_score[features].copy()
X

,INCOME,DEBT,R_EXPENDITURE,R_ENTERTAINMENT,CAT_GAMBLING
0,33269,532304,0.5797,0.5065,High
1,77158,315648,0.3580,0.5042,No
2,30917,534864,0.5698,0.6178,High
3,80657,629125,0.4985,0.5094,High
4,149971,2399531,0.4604,0.5150,High
...,...,...,...,...,...
995,328892,5501471,0.5193,0.5038,High
996,81404,680837,0.4252,0.6379,No
997,0,30760,0.7427,0.6163,No
998,36011,604181,0.6502,0.5105,No


In [10]:
X.describe()

,INCOME,DEBT,R_EXPENDITURE,R_ENTERTAINMENT
count,1000.000000,1.000000e+03,1000.000000,1000.000000
mean,121610.019000,7.907180e+05,0.512560,0.546432
std,113716.699591,9.817904e+05,0.079740,0.062581
min,0.000000,0.000000e+00,0.176900,0.389800
25%,30450.250000,5.396675e+04,0.469675,0.507875
50%,85090.000000,3.950955e+05,0.519900,0.511900
75%,181217.500000,1.193230e+06,0.561150,0.590800
max,662094.000000,5.968620e+06,0.742700,0.787600


In [ ]:
X['GAMBLING_LOW'] = X['CAT_GAMBLING'].apply(lambda x: 1 if x == 'Low' else 0)
X['GAMBLING_HIGH'] = X['CAT_GAMBLING'].apply(lambda x: 1 if x == 'High' else 0)

Tasks for User Story: 

Conduct a literature review on XAI techniques and compile relevant methods 

Evaluate the applicability of each identified XAI technique to the current project and document findings 

Select 3 XAI techniques from the researched list for implementation 

Integrate the selected XAI techniques into the existing model pipeline 

Tasks for User Story 4.2: 

Implement at least 3 model explainability techniques. 

Perform hyperparameter tuning to optimise model performance 

Experiment with various feature engineering and feature selection methods 

Apply techniques to handle dataset imbalance and assess their impact 

Report what combination of techniques, model and explainability method worked so far.  

Achieve a minimum F1 score of 65%  